### PR TITLE
[unittests] Add a matrix recognition test

### PR DESCRIPTION
**Pull request by qcolombet**
The main characteristic of the test is that it features two inputs.
The goal is to make sure the scheduler (and graph lowering in general)
is able to properly enforce the dependencies between the nodes with
something more complicated than a straight chain of operators.

### CODE REVIEW

**Comment by qcolombet** [comment]
Note: I could completely get rid of the templatization of the Matrix3x3 struct and such. The reason I added it in the first place was because I wanted to have a network with FloatTy and one with Int8Ty.
I realized after the fact that Int8Ty was Int8QTy and cannot be used directly (i.e., it needs to go through quantization).

**Comment by qcolombet** [comment]
Another note: I added the test as part of the JIT test, but maybe we want it to be standalone.

**Comment by nadavrot** [comment]
Quentin, thanks for working on this. 

The inc file is pretty big. Could you generate this data programmatically?   

This is a pretty big test for a relatively small neural network. For comparison, the learnXor unit test is ~60 lines of code. Would it be possible to simplify the test and fit it into the MLTest framework? The MLTest unit test runs the code on the three backends. We'll need to tune the parameters to make sure that the test finishes training in a reasonable amount of time, even in debug mode.

**Comment by qcolombet** [comment]
> The inc file is pretty big. Could you generate this data programmatically?

The inc file is actually already generated programmatically (see generateTrainingDataFor).
The reasons I saved one of this run into a file were:
- It seemed wasteful to generate new data each time we run the test (+ it makes debugging a little bit harder where one needs to provide the seeds and so on.)
- I wanted the performance of the prediction to be controlled and thus, having a fixed input helped in that sense.

Let me know if you prefer still we generate the data set at runtime and I'll do the changes.

> Would it be possible to simplify the test and fit it into the MLTest framework?

Simplify in that sense?

As of fitting it into the MLTest framework, I'll have a look.

> We'll need to tune the parameters to make sure that the test finishes training in a reasonable amount of time, even in debug mode.

On my machine, the whole JITTest run in ~20 sec in debug mode (unless I misconfigured something).
The added matrix tests alone are about 6 sec total.

Is that reasonable?

File: tests/unittests/MatrixRotationTest.cpp [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+  // changed.
+  ASSERT_EQ(numSamples, 3191);
+  ASSERT_EQ(numMatrices, 1594);
+  testNetwork<float>(trainingDataForFloat, numMatrices,
+                     trainingDataForFloatRelationshipData, numSamples,
+                     /*learningRate*/ 0.02, /*batchSize*/ 17,
+                     /*iterations*/ 27000, /*errors*/ 0);
+}
+
+TEST(MatrixRotationTest, networkInt8) {
```
Comment (rdzhabarov) [inline tests/unittests/MatrixRotationTest.cpp]: "networkInt8" what is the goal here?

The network itself is not the quantized network. I do not think this test brings a lot of value, but makes a lot of things to be templatized (which might not be necessary).

Comment (qcolombet) [reply tests/unittests/MatrixRotationTest.cpp]: Agreed.

That's what I noted in some earlier post:
> Note: I could completely get rid of the templatization of the Matrix3x3 struct and such. The reason I added it in the first place was because I wanted to have a network with FloatTy and one with Int8Ty.
> I realized after the fact that Int8Ty was Int8QTy and cannot be used directly (i.e., it needs to go through quantization).

File: tests/unittests/MatrixRotationTest.cpp [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+
+#include <cstddef> // for size_t.
+#include <cstdint> // for uint8_t.
+#include <utility> // for std::pair.
+
+using namespace glow;
+
+namespace {
+template <typename Element> struct Matrix3x3 {
+  Element m_[3][3];
```
Comment (rdzhabarov) [inline tests/unittests/MatrixRotationTest.cpp]: maybe just std::array of std::array.  ctor and operator== will be trivial in that case.

**Comment by nadavrot** [comment]
@qcolombet We have a few kinds of tests in the Glow repo right now. I envision the rotation test as something that should go into the unit tests, which means that it should run on every commit. Similar to make-check in LLVM. Similar test in (learnXor, circleTest, linear regression test) run within 2-3 seconds in debug mode. It's possible to reduce the runtime by tuning the parameters. For example, you are now running 30,000 trining iterations. Maybe it's possible to increase the learning rate and reduce the number of iterations? Maybe it's possible to reduce the size of the fullyconnected layers to reduce the size of the matrix multipliers?

File: tests/unittests/MatrixRotationTest.cpp (hunk 1 above)
Comment (rdzhabarov) [reply tests/unittests/MatrixRotationTest.cpp]: nice!

**Comment by nadavrot** [comment]
Here are a few ideas of how to simplify the test:
1. The code does not need to be templated on the element type. Neural networks are always trained in floating point mode. When we perform quantization on the network the conversion is automatic and the interfaces remain floating-point tensors. 
2. Instead of defining matrix3x3 you could use Tensor with the right dimensions. Tensor already has methods for dumping the ascii representation of the tensor (we use this in the circle unit test to visualize the region). 
3. Instead of creating unowned tensors you could use the runBatch api that knows how to copy data into some slices. 
4. nextRand is stable, in the sense that it always returns the same random numbers on all platforms on all tests. It's not stable when you change the order of tests, but @stoklund had a suggestion to seed the random number generator before every test execution. This means that you don't need to generate the training data yourself in a large 'inc' fine. Also, the neural network should be able to handle noise and learn a simple problem such as this one even with random inputs.

**Comment by qcolombet** [comment]
> The code does not need to be templated on the element type. Neural networks are always trained in floating point mode. When we perform quantization on the network the conversion is automatic and the interfaces remain floating-point tensors.

Agreed, already noted that in an earlier post. Current version is out of template.

> Instead of defining matrix3x3 you could use Tensor with the right dimensions. Tensor already has methods for dumping the ascii representation of the tensor (we use this in the circle unit test to visualize the region).

Taking a look.

> Instead of creating unowned tensors you could use the runBatch api that knows how to copy data into some slices.

I thought runBatch was just for training and I used it for that purpose, but I thought we had to do the slicing by hand when running in inference mode. This is what we do in HyphenTest for instance.
Am I missing something?

> [...]This means that you don't need to generate the training data yourself in a large 'inc' fine.

That still means we will pay the cost for generating the training data every single time, which I thought was going against the fast llvm-check style that we want for every commit.
I am going to reduce drastically the size of the input and just embedded that so that we have the best of both worlds. My concern was this may not be enough data for the network to realize what it needs to recognize.

> Also, the neural network should be able to handle noise and learn a simple problem such as this one even with random inputs.

To me the point was not to test the NN, but really what Glow does with the NN. What I am saying is that I was not concerned that this NN only works well with this training set and doesn't handle well the noise.

What do you think?

File: tests/unittests/MatrixRotationTest.cpp [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+    out << "  {" << rel.matrixIdxA_ << ',' << rel.matrixIdxB_ << ','
+        << rel.clockwiseRotationStepsAToB_ << "},\n";
+  }
+
+  out << "}; // End of " << relationshipVarName << ".\n";
+}
+
+
+// --------------------- Tests for Matrix3x3 start here ----------------------
+static float m1[3][3] = {
```
Comment (rdzhabarov) [inline tests/unittests/MatrixRotationTest.cpp]: nit: you can make all those std::array and make ctor ```Matrix3x3(const float m[3][3]) ``` to get an array as input param.

**Comment by qcolombet** [comment]
I generated a much smaller data set and embedded it in the file.
The test runs for about 1s now.
I still need to:
- Use the tensor representation directly
- Move the thing in MLTest
- Fix std::array nits

**Comment by qcolombet** [comment]
For the record, reducing the number of FC layers doesn't help that much the execution time but affected the performance of the recognition.
Anyhow right now the test has only 60 iterations, so that shouldn't matter.

**Comment by nadavrot** [comment]
@qcolombet Unit tests have two parts: the machine learning neural network, and the story that wraps the network. For example, we have a test that uses FC and Softmax to classify basketball and football players. The FC and Softmax are the network and the stuff related to sports is the story around the test. 

We'd like to maximize the ML part and minimize the story part. The ML in this pull request is great, but it comes with 600 lines of code around it, that does not add much safety to the program. The tests in MLTest.cpp are a good example of concise tests. Te Xor test is 60 lines of code. It generates data initializes the network and checks the result in a few lines of code. The code is very readable and easy to debug. 

The learnXor test uses the MLTest utilities so it automatically runs on 3 backend. And when we add new backends or improve the infrastructure the test will gain from it automatically. 

Please simplify the test and integrate it into MLTest. Take a look at the other tests and try to fit the code into ~100 loc. It should be relatively easy and fast to generate data (randomize a tensor and rotate it). Verification is also very easy since you are generating the data.

**Comment by qcolombet** [comment]
Thanks @nadavrot for the taking the time of describing the rationale.
I get the point now.
Anyway, my plan was to look into MLTest.

**Comment by qcolombet** [comment]
Refactored the patch to be a subtest in the MLTest "group".

The patch now consists in two main function:
- One function generate the data at runtime.
- One function create the network, train it with the random data, and check that we overfitted the data.
The test takes about 1 sec with the interpreter. (This includes generating the data.)

Thanks to @artemrakhov for suggesting to use only 0s and 1s in the input matrices.

File: tests/unittests/MLTest.cpp [hunk 4]
```diff
@@ -983,6 +983,204 @@ TEST_P(InterpreterAndCPU, testFindPixelRegression) {
...
+      // Fetch the cell registered in the clockwiseOrder at the desired step.
+      for (size_t i = 0; i != possibleTargetCells; ++i) {
+        const CellIdx &sourceCellIdx = clockwiseOrder[i];
+        const CellIdx &targetCellIdx =
+            clockwiseOrder[(i + clockwiseSteps) % possibleTargetCells];
+        handleMatricesB.at({idx, targetCellIdx.first, targetCellIdx.second}) =
+            handleMatricesA.at({idx, sourceCellIdx.first, sourceCellIdx.second});
+      }
+    } else {
+      // There is a high probability that A and B don't have a rotation
```
Comment (nadavrot) [inline tests/unittests/MLTest.cpp]: I am not sure why we need this part. Even if our matrices support only boolean values, the probability of randomly selecting a matrix that happen to be a rotation of another matrix is very low.  (2^9 / 8), which is way below the accuracy that we expect from a network (90% accuracy is fine). If we want to be extra conservative we can select the input values -1, 0, 1, and this will reduce the probability of accidentally generating a random matrix significantly. The neural network should learn to generalize and ignore these inputs. In fact, introducing noise is a way to make neural networks to generalize. So, I don't think that accidentally generating the wrong label in a very small number of cases is something that we should worry about.

File: tests/unittests/MLTest.cpp [hunk 5]
```diff
@@ -983,6 +983,204 @@ TEST_P(InterpreterAndCPU, testFindPixelRegression) {
...
+      mod.createVariable(ElemKind::FloatTy, {config.batchSize, 3, 3}, "matrixB",
+                         VisibilityKind::Public, Variable::TrainKind::None);
+  Variable *varExpected =
+      mod.createVariable(ElemKind::IndexTy, {config.batchSize, 1}, "expected",
+                         VisibilityKind::Public, Variable::TrainKind::None);
+
+  // Simply concatenating the matrices first would probability be as effective
+  // but we want to build something more complex than a straight chain of
+  // operators to stress the scheduler.
+  auto *FCA = F->createFullyConnected("hidden_matrixA_fc", varMatricesA, 10);
```
Comment (nadavrot) [inline tests/unittests/MLTest.cpp]: The network looks really good.

File: tests/unittests/MLTest.cpp [hunk 6]
```diff
@@ -983,6 +983,204 @@ TEST_P(InterpreterAndCPU, testFindPixelRegression) {
...
+         "Size of the tensors is incompatible");
+  auto handleMatricesA = matricesA.getHandle<float>();
+  auto handleMatricesB = matricesB.getHandle<float>();
+  auto handleExpected = expected.getHandle<size_t>();
+
+  handleMatricesA.randomize(0.0, 2.0, PRNG);
+  handleMatricesB.randomize(0.0, 2.0, PRNG);
+  for (unsigned idx = 0; idx < numSamples; ++idx) {
+    // Narrow the search space for the network by using only 0s and 1s.
+    for (size_t rowIdx = 0; rowIdx != 3; ++rowIdx) {
```
[the remark this answers is not in the document]
Comment (rdzhabarov) [reply tests/unittests/MLTest.cpp]: iirc, randomize has an overload with ints, randomize(0, 1) and we'd not need to do any additional checks.

[the remark this answers is not in the document]
Comment (qcolombet) [reply tests/unittests/MLTest.cpp]: Thanks @rdzhabarov, I haven't realized we had the ints randomize method.

**Comment by nadavrot** [comment]
great!
